orcid-core/src/main/java/org/orcid/core/manager/impl/OrcidProfileManagerImpl.java

```diff
@@ -382,50 +382,63 @@ public OrcidProfile createOrcidProfileAndNotify(OrcidProfile orcidProfile) {
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                ProfileEntity existingProfileEntity = profileDao.find(orcidProfile.getOrcidIdentifier().getPath());
+                Visibility defaultVisibility = Visibility.fromValue(existingProfileEntity.getActivitiesVisibilityDefault().value());
+
+                if (existingProfileEntity != null) {
+                    // Dont delete the existing elements anymore
+                    // profileDao.removeChildrenWithGeneratedIds(existingProfileEntity);
+                    setWorkPrivacy(orcidProfile, defaultVisibility);
+                    setAffiliationPrivacy(orcidProfile, defaultVisibility);
+                    setFundingPrivacy(orcidProfile, defaultVisibility);
+                }
+                dedupeWorks(orcidProfile);
+                dedupeAffiliations(orcidProfile);
+                dedupeFundings(orcidProfile);
+                addSourceToEmails(orcidProfile, existingProfileEntity, amenderOrcid);
 
-        ProfileEntity profileEntity = adapter.toProfileEntity(orcidProfile, existingProfileEntity, updateOptions);
-        profileEntity.setLastModified(new Date());
-        profileEntity.setIndexingStatus(IndexingStatus.PENDING);
-        profileDao.flush();
-        ProfileEntity updatedProfileEntity = profileDao.merge(profileEntity);
-        profileDao.refresh(updatedProfileEntity);
-        OrcidProfile updatedOrcidProfile = convertToOrcidProfile(updatedProfileEntity, LoadOptions.ALL);
+                Boolean claimed = orcidProfile.getOrcidHistory() != null ? orcidProfile.getOrcidHistory().isClaimed() : existingProfileEntity.getClaimed();
 
-        orcidProfileCacheManager.put(updatedOrcidProfile);
-        return updatedOrcidProfile;
+                if (orcidProfile.getOrcidInternal() != null && orcidProfile.getOrcidInternal().getPreferences() != null
+                        && orcidProfile.getOrcidInternal().getPreferences().getActivitiesVisibilityDefault() != null) {
+                    defaultVisibility = orcidProfile.getOrcidInternal().getPreferences().getActivitiesVisibilityDefault().getValue();
+                }
+                addDefaultVisibilityToBioItems(orcidProfile, defaultVisibility, claimed);
+
+                ProfileEntity profileEntity = adapter.toProfileEntity(orcidProfile, existingProfileEntity, updateOptions);
+                profileEntity.setLastModified(new Date());
+                profileEntity.setIndexingStatus(IndexingStatus.PENDING);
+                profileDao.flush();
+                ProfileEntity updatedProfileEntity = profileDao.merge(profileEntity);
+                profileDao.refresh(updatedProfileEntity);
+                
+                // Then persist the works
+                if(orcidProfile.getOrcidActivities() != null && orcidProfile.getOrcidActivities().getOrcidWorks() != null) {
```

Comment:
@wjrsimpson: And then we call adapter.setWorks which will iterate over the list of works and create/update/delete as needed.

After this, the transaction is closed, so, everything get stored in the DB at the same time